File: src/main/java/mandala/WorldPathNN.java
// For conditions of distribution and use, see copyright notice in Mandala.java

// World path neural network.

package mandala;

import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;

import org.nd4j.linalg.api.ndarray.INDArray;

public class WorldPathNN
{
	// Dataset file name.
	public static String DATASET_FILE_NAME = "world_path_nn_dataset.csv";
	
	// Results file name.
	public static String RESULTS_FILE_NAME = "world_path_nn_results.json";
	
	// Dimensions of code and hidden vectors.
	public static int CODE_DIMENSION = 384;
	public static int HIDDEN_DIMENSION = 384;
	
    // Usage.
    public static final String Usage =
      "Usage:\n" +
      "    java mandala.WorldPathNN\n" +
      "      [-datasetFilename <file name> (default=" + DATASET_FILE_NAME + ")]\n" +     		  
      "      [-codeDimension <quantity> (default=" + CODE_DIMENSION + ")]\n" +
      "      [-hiddenDimension <quantity> (default=" + HIDDEN_DIMENSION + ")]\n" +
      "      [-resultsFilename <file name> (default=" + RESULTS_FILE_NAME + ")]\n" +      
      "Exit codes:\n" +
      "  0=success\n" +
      "  1=error";   
	
    // Main.
    public static void main(String[] args) 
    {
        for (int i = 0; i < args.length; i++)
        {
           if (args[i].equals("-datasetFilename"))
           {
              DATASET_FILE_NAME = args[i];
              continue;
           }       
           if (args[i].equals("-codeDimension"))
           {
              i++;
              if (i >= args.length)
              {
                 System.err.println("Invalid codeDimension option");
                 System.err.println(Usage);
                 System.exit(1);
              }
              try
              {
                 CODE_DIMENSION = Integer.parseInt(args[i]);
              }
              catch (NumberFormatException e) {
                 System.err.println("Invalid codeDimension option");
                 System.err.println(Usage);
                 System.exit(1);
              }
              if (CODE_DIMENSION < 0)
              {
                 System.err.println("Invalid codeDimension option");
                 System.err.println(Usage);
                 System.exit(1);
              }              
              continue;
           }
           if (args[i].equals("-hiddenDimension"))
           {
              i++;
              if (i >= args.length)
              {
                 System.err.println("Invalid hiddenDimension option");
                 System.err.println(Usage);
                 System.exit(1);
              }
              try
              {
            	  HIDDEN_DIMENSION = Integer.parseInt(args[i]);
              }
              catch (NumberFormatException e) {
                 System.err.println("Invalid hiddenDimension option");
                 System.err.println(Usage);
                 System.exit(1);
              }
              if (HIDDEN_DIMENSION < 0)
              {
                 System.err.println("Invalid hiddenDimension option");
                 System.err.println(Usage);
                 System.exit(1);
              }              
              continue;
           }
           if (args[i].equals("-resultsFilename"))
           {
              RESULTS_FILE_NAME = args[i];
              continue;
           }               
           if (args[i].equals("-help"))
           {
                 System.out.println(Usage);
                 System.exit(0);
           }                        
           System.err.println("Invalid option: " + args[i]);
           System.err.println(Usage);
           System.exit(1);
        }
        
        // Create NN.
        MandalaNN.CODE_DIMENSION = CODE_DIMENSION;
        MandalaNN.HIDDEN_DIMENSION = HIDDEN_DIMENSION;
        MandalaNN mandalaNN = new MandalaNN();
        
        // Import dataset file.
        mandalaNN.importDataset(DATASET_FILE_NAME);
        
        // Build NN.
        mandalaNN.build();
        
        // Train.
        mandalaNN.train(1000);
        
        // Validate.
        mandalaNN.validate();
                
        // Test.
        mandalaNN.test();
        
        // Results.
        INDArray effects = mandalaNN.trainEffectData;
        INDArray predictions = mandalaNN.trainPredictions;
        int trainErrors = 0;
        int trainTotal = 0;
        for (int i = 0, j = predictions.rows(); i < j; i++)
        { 
    		if (maxValIdx(effects.getRow(i)) != maxValIdx(predictions.getRow(i)))
			{
    			trainErrors++;			
			}
    		trainTotal++;     		
        }
        System.out.print("Train prediction errors/total = " + trainErrors + "/" + trainTotal);
        float trainErrorPct = 0.0f;
        if (trainTotal > 0)
        {
            trainErrorPct = ((float)trainErrors / (float)trainTotal) * 100.0f;
            System.out.print(" (");
            System.out.printf("%.2f", trainErrorPct);
            System.out.print("%)"); 
        }
        System.out.println();
        effects = mandalaNN.testEffectData;
        predictions = mandalaNN.testPredictions;
        int testErrors = 0;
        int testTotal = 0;
        for (int i = 0, j = predictions.rows(); i < j; i++)
        { 
    		if (maxValIdx(effects.getRow(i)) != maxValIdx(predictions.getRow(i)))
			{
    			testErrors++;			
			}
    		testTotal++;     		
        }
        System.out.print("Test prediction errors/total = " + testErrors + "/" + testTotal);
        float testErrorPct = 0.0f;
        if (testTotal > 0)
        {
            testErrorPct = ((float)testErrors / (float)testTotal) * 100.0f;
            System.out.print(" (");
            System.out.printf("%.2f", testErrorPct);
            System.out.print("%)"); 
        }
        System.out.println();        
		try
		{
	        FileWriter fileWriter = new FileWriter(RESULTS_FILE_NAME);
	        PrintWriter printWriter = new PrintWriter(fileWriter);
	        printWriter.write("{");        	        
	        printWriter.write("\"train_prediction_errors\":\"" + trainErrors + "\",");	        
	        printWriter.write("\"train_total_predictions\":\"" + trainTotal + "\",");	  	        
	        printWriter.write("\"train_error_pct\":\"" + trainErrorPct + "\",");	        	        	        
	        printWriter.write("\"test_prediction_errors\":\"" + testErrors + "\",");	        
	        printWriter.write("\"test_total_predictions\":\"" + testTotal + "\",");	  	        
	        printWriter.write("\"test_error_pct\":\"" + testErrorPct + "\"");	        	        
	        printWriter.write("}\n");    
	        printWriter.close();
		} catch (IOException e)
		{
			System.err.println("Cannot write results to file " + RESULTS_FILE_NAME);
			System.exit(1);
		}
		
        System.exit(0);
    }
    
    // Get index of max value in row.
    public static int maxValIdx(INDArray row)
    {
    	if (row.length() == 0)
    	{
    		System.err.println("maxValIdx: empty row");
    		System.exit(1);
    	}
    	int maxIdx = 0;
    	float maxVal = row.getFloat(0);
    	for (int i = 1; i < row.length(); i++)
    	{
    		float val = row.getFloat(i);
    		if (val > maxVal)
    		{
    			maxIdx = i;
    			maxVal = val;
    		}
    	}
    	return maxIdx;    	
    }
}

File: src/main/python/world_path_nn.py
# For conditions of distribution and use, see copyright notice in Mandala.java
# World path NN.

from numpy import array, argmax
from numpy import loadtxt
from keras.models import Sequential
from keras.layers import Dense
import sys, getopt

# parameters
n_neurons = 128
n_epochs = 150

# results file name
results_filename = 'world_path_nn_results.json'

# get options
try:
  opts, args = getopt.getopt(sys.argv[1:],"hn:e:",["neurons=","epochs="])
except getopt.GetoptError:
  print('world_path_nn.py [-n <hidden_neurons>] [-e <epochs>]')
  sys.exit(1)
for opt, arg in opts:
  if opt == '-h':
     print('world_path_nn.py [-n <hidden_neurons>] [-e <epochs>]')
     sys.exit(0)
  if opt in ("-n", "--neurons"):
     n_neurons = int(arg)
  elif opt in ("-e", "--epochs"):
     n_epochs = int(arg)
  else:
     print('world_path_nn.py [-n <hidden_neurons>] [-e <epochs>]')
     sys.exit(1)

# import dataset
from world_path_nn_dataset import X_train_shape, y_train_shape, X_train, y_train, X_test_shape, y_test_shape, X_test, y_test
if X_train_shape[0] == 0:
    print('Empty train dataset')
    sys.exit(1)
if X_test_shape[0] == 0:
    print('Empty test dataset')
    sys.exit(1)
seq = array(X_train)
X = seq.reshape(X_train_shape[0], X_train_shape[1])
seq = array(y_train)
y = seq.reshape(y_train_shape[0], y_train_shape[1])

# create NN
model = Sequential()
model.add(Dense(n_neurons, input_dim=X_train_shape[1], activation='relu'))
model.add(Dense(n_neurons, activation='relu'))
model.add(Dense(y_train_shape[1], activation='sigmoid'))
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

# train
model.fit(X, y, epochs=n_epochs, batch_size=10)
_, accuracy = model.evaluate(X, y)
print('Accuracy: %.2f' % (accuracy*100))

# validate
seq = array(X_train)
X = seq.reshape(X_train_shape[0], X_train_shape[1])
seq = array(y_train)
y = seq.reshape(y_train_shape[0], y_train_shape[1])
predictions = model.predict(X, batch_size=X_train_shape[0], verbose=2)
trainOK = 0
trainErrors = 0
trainTotal = 0
for path in range(X_train_shape[0]):
    p = []
    for step in range(y_train_shape[1]):
        r = argmax(predictions[path][step])
        p.append(r)
    t = []
    for step in range(y_train_shape[1]):
        r = argmax(y[path][step])
        t.append(r)
    if p == t:
        trainOK += 1
    else:
        errs = 0
        for i in range(len(p)):
            if p[i] != t[i]:
                errs += 1
        trainErrors += errs
    trainTotal += len(p)

# predict
seq = array(X_test)
X = seq.reshape(X_test_shape[0], X_test_shape[1])
seq = array(y_test)
y = seq.reshape(y_test_shape[0], y_test_shape[1])
predictions = model.predict(X, batch_size=X_test_shape[0], verbose=2)
testOK = 0
testErrors = 0
testTotal = 0
for path in range(X_test_shape[0]):
    p = []
    for step in range(y_test_shape[1]):
        r = argmax(predictions[path][step])
        p.append(r)
    t = []
    for step in range(y_test_shape[1]):
        r = argmax(y[path][step])
        t.append(r)
    if p == t:
        testOK += 1
    else:
        errs = 0
        for i in range(len(p)):
            if p[i] != t[i]:
                errs += 1
        testErrors += errs
    testTotal += len(p)

# Print results.
print("Train correct paths/total = ", trainOK, "/", X_train_shape[0], sep='', end='')
if X_train_shape[0] > 0:
    r = (float(trainOK) / float(X_train_shape[0])) * 100.0
    print(" (", str(round(r, 2)), "%)", sep='', end='')
print(", prediction errors/total = ", trainErrors, "/", trainTotal, sep='', end='')
trainErrorPct=0
if trainTotal > 0:
    trainErrorPct = (float(trainErrors) / float(trainTotal)) * 100.0
    print(" (", str(round(trainErrorPct, 2)), "%)", sep='', end='')
print('')
print("Test correct paths/total = ", testOK, "/", X_test_shape[0], sep='', end='')
if X_test_shape[0] > 0:
    r = (float(testOK) / float(X_test_shape[0])) * 100.0
    print(" (", str(round(r, 2)), "%)", sep='', end='')
print(", prediction errors/total = ", testErrors, "/", testTotal, sep='', end='')
testErrorPct=0
if testTotal > 0:
    testErrorPct = (float(testErrors) / float(testTotal)) * 100.0
    print(" (", str(round(testErrorPct, 2)), "%)", sep='', end='')
print('')

# Write results to file.
with open(results_filename, 'w') as f:
    f.write('{')
    f.write('\"train_correct_paths\":\"'+str(trainOK)+'\",')
    f.write('\"train_total_paths\":\"'+str(X_train_shape[0])+'\",')
    f.write('\"train_prediction_errors\":\"'+str(trainErrors)+'\",')
    f.write('\"train_total_predictions\":\"'+str(trainTotal)+'\",')
    f.write('\"train_error_pct\":\"'+str(round(trainErrorPct, 2))+'\",')
    f.write('\"test_correct_paths\":\"'+str(testOK)+'\",')
    f.write('\"test_total_paths\":\"'+str(X_test_shape[0])+'\",')
    f.write('\"test_prediction_errors\":\"'+str(testErrors)+'\",')
    f.write('\"test_total_predictions\":\"'+str(testTotal)+'\",')
    f.write('\"test_error_pct\":\"'+str(round(testErrorPct, 2))+'\"')
    f.write('}\n')


sys.exit(0)
